feat: FAQ Entry: git dependencies on closed networks

mahikhond opened the pull request:
Details:
- FAQ Entry: what should users do with git dependencies on closed networks

mkhond committed:
feat: FAQ Entry: git dependencies on closed networks

Details:
- FAQ Entry: what should users do with git dependencies on closed networks

bcoe commented on troubleshooting/faq.md:
```diff
@@ -161,6 +162,26 @@ So it’s important to understand the different storage driver options available
 
+## What should I do with git dependencies on closed networks?
+
+If you are installing the package that have a dependency on a git url, npm will fetch that package from github itself because git dependencies are valid in npm.
```
a couple edits:

rather than:

`If you are installing the package that have a dependency on a git url`

let's go with:


`If you are installing a package that has a git URL as a dependency, npm will fetch that package from the remote git repository itself.`

rather than

`npm will fetch that package from github itself because git dependencies are valid in npm.`

Let's go with:

`. npm supports both git and HTTP URL formats (see https://docs.npmjs.com/files/package.json#dependencies), so we will need to work around this in our private registry.`

bcoe commented on troubleshooting/faq.md:
````diff
@@ -161,6 +162,26 @@ So it’s important to understand the different storage driver options available
+  "name2" : "git://github.com/user/project.git#commit-ish"
+}
+```
+When you are installing the package in the closed network where a dependency references a git url rather than a package name, you are unable to access due to no internet connectivity.
````
`you are unable to access due to no internet connectivity`

->

`you will be unable to access due to no internet connectivity`

bcoe commented on troubleshooting/faq.md:
````diff
@@ -161,6 +162,26 @@ So it’s important to understand the different storage driver options available
+```
+When you are installing the package in the closed network where a dependency references a git url rather than a package name, you are unable to access due to no internet connectivity.
+
+In this situation you have to download the tarball of package from github on a local and re-publish to the private registry.
````
`In this situation, one approach you can take is to download a tarball version of the package and republish it to your private registry.`

Let's add a note about fetching the tarball version of a tag on GitHub, documentation for doing this can be found here, which we can work from:

http://stackoverflow.com/questions/5746325/how-do-i-download-a-tarball-from-github-using-curl

bcoe reviewed (changes requested):
@mahikhond a few edits for you; this is sounding like a great start.

mkhond committed:
fix: Updated the requested changes.

Details:
- Updated the requested changes regarding the tarball download from github.

bcoe commented on troubleshooting/faq.md:
```diff
@@ -161,6 +162,33 @@ So it’s important to understand the different storage driver options available
 
+## What should I do with git dependencies on closed networks?
+
+If you are installing a package that has a git URL as a dependency, npm will fetch that package from the remote git repository itself. npm supports both git and HTTP URL formats (see https://docs.npmjs.com/files/package.json#dependencies), so we will need to work around this in our private registry.
```
let's make this URL a markdown URL.

bcoe commented on troubleshooting/faq.md:
````diff
@@ -161,6 +162,33 @@ So it’s important to understand the different storage driver options available
+You can download a tarball version of the package from GitHub using
+```
+ wget -L https://github.com/user-or-org/repo/archive/master.tar.gz
+
````
extra space here.

bcoe reviewed (changes requested):
sounds good, just a couple more tiny edits.

bcoe commented on troubleshooting/faq.md:
```diff
@@ -161,6 +162,33 @@ So it’s important to understand the different storage driver options available
+
+*  tar -xvzf package.tar.gz
+*  cd package
+*  npm publish --registry=http://your-private-registry:8080
```
thinking about it, we should probably add one more section that describes how to use the new dependency;

People will need to switch the packages referencing the git URL, to reference the dependency that is now published to their private registry.

Perhaps we can split this into two sections:

`publishing the git dependency to your private repo`

and

`updating packages to reference the new dependency`

We should note that, since it will often be a dependency of a dependency that has the git dependency, it will likely be necessary to both publish the git dependency and the dependency requiring it. A good best practice is to publish these new versions with a scope (@mycompany/request);

we have a tool for simplifying this process: https://www.npmjs.com/package/scope-it

mkhond committed:
fix: Updated the requested changes.

Details:
- Added the two separate section publishing the git dependency to private repo and updating packages to reference the new dependency.

arobson commented on troubleshooting/faq.md:
````diff
@@ -161,6 +162,47 @@ So it’s important to understand the different storage driver options available
+  "name2" : "git://github.com/user/project.git#commit-ish"
+}
+```
+When you are installing the package in the closed network where a dependency references a git url rather than a package name, you will be unable to access due to no internet connectivity.
````
Minor change: "you will be unable to install the dependency since it references an external network address." instead of "you will be unable to access due to no internet connectivity."

arobson commented on troubleshooting/faq.md:
````diff
@@ -161,6 +162,47 @@ So it’s important to understand the different storage driver options available
+```
+Replace `user-or-org` and `repo` accordingly.
+
+To publish the tarball to private registry simply follow the steps listed below:
````
Minor change: "To publish the tarball to your private registry simply follow the steps listed below:"

arobson commented on troubleshooting/faq.md:
```diff
@@ -161,6 +162,47 @@ So it’s important to understand the different storage driver options available
+
+In this situation, one approach you can take is to download a tarball version of the package and republish it to your private registry.
+
+You can download a tarball version of the package from GitHub using
```
Minor change - just add a `:` to the end of the line for consistency please :)

mkhond committed:
fix: Updated suggested minor changes.

mkhond committed:
fix: Removed extra line

arobson commented:
@mahikhond - once you have a chance to rebase this, we should be able to merge this in and get it published.

mahikhond committed:
Merge branch 'master' into git-dependencies-faq
